File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        for i in s:
            if i=='(' or i=='[' or i=='{':
                stack.append(i)
            else:
                if not stack:#if stack is empty and that means () is over so for [] to start there will be [ already in stack
                    return False
                top=stack.pop()
                if i==')' and top!='(':
                    return False
                elif i==']' and top!='[':
                    return False
                elif i=='}' and top!='{':
                    return False
        if len(stack)==0:
            return True
        return False

                

        

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        boolean ans=false;
        for (int i=0;i<s.length();i++){
            if(s.charAt(i)=='(' || s.charAt(i)=='{' || s.charAt(i)=='['){
                stack.push(s.charAt(i));
            }
            else{
                if(!stack.isEmpty()){
                char c=stack.pop();
                
                if(c=='(' && s.charAt(i)==')'){
                    ans=true;
                    continue;
                }
                if(c=='[' && s.charAt(i)==']'){
                    ans=true;
                    continue;

                }
                if(c=='{' && s.charAt(i)=='}'){
                    ans=true;
                    continue;
                }
                }
            ans=false;
            return ans;
                
            }
        }
        if(!stack.isEmpty()){
            return false;
        }
        return ans;

    }
}